optaplanner-examples/src/main/java/org/optaplanner/examples/batchscheduling/solver/score/BatchSchedulingEasyScoreCalculator.java

```diff
@@ -0,0 +1,244 @@
+
+import org.optaplanner.core.api.score.buildin.bendablelong.BendableLongScore;
+import org.optaplanner.core.api.score.calculator.EasyScoreCalculator;
+import org.optaplanner.examples.batchscheduling.domain.Allocation;
+import org.optaplanner.examples.batchscheduling.domain.AllocationPath;
+import org.optaplanner.examples.batchscheduling.domain.Schedule;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class BatchSchedulingEasyScoreCalculator implements EasyScoreCalculator<Schedule, BendableLongScore> {
+
+    final Logger logger = LoggerFactory.getLogger(BatchSchedulingEasyScoreCalculator.class);
+    public static final String ROUTE_PATH_SEPERATOR = "---";
+    private static final int OTHER_ALLOCATION_PENALTY = 1;
+    private static final int CURRENT_ALLOCATION_PENALTY = 1;
+    private static final boolean COMPUTE_hard1Score = true;
+    private static final boolean COMPUTE_soft2Score = true;
+
+    public BendableLongScore calculateScore(Schedule schedule) {
+
+        //Refer to IncrementalScoreCalculator for description 
+        long hard0Score = 0;
+        long hard1Score = 0;
+        long hard2Score = 0;
+        long soft0Score = 0;
+        long soft1Score = 0;
+        long soft2Score = 0;
+
+        Map<Long, String> batchRoutePathMap = new HashMap<Long, String>();
+        Map<Long, Long> batchOtherPenaltyValueMap = new HashMap<Long, Long>();
+        Map<Long, Long> batchCurrentPenaltyValueMap = new HashMap<Long, Long>();
+        Map<String, Long> segmentOverlapMap = new HashMap<String, Long>();
+
+        if (schedule.getAllocationPathList() != null) {
+            for (AllocationPath allocationPath : schedule.getAllocationPathList()) {
+                if (allocationPath.getRoutePath() != null) {
+                    batchRoutePathMap.put(allocationPath.getBatch().getId(), allocationPath.getRoutePath().getPath());
+
+                    String[] array1 = getSegmentArray(allocationPath.getRoutePath().getPath());
+
+                    for (AllocationPath allocationPath2 : schedule.getAllocationPathList()) {
+                        if ((allocationPath2.getRoutePath() != null)
+                                && (!(allocationPath.getBatch().equals(allocationPath2.getBatch())))) {
+                            for (String str2 : getSegmentArray(allocationPath2.getRoutePath().getPath())) {
+                                for (String str1 : array1) {
+                                    if (str2.equals(str1)) {
+                                        soft2Score += 1;
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+
+                batchOtherPenaltyValueMap.put(allocationPath.getBatch().getId(), 0L);
+                batchCurrentPenaltyValueMap.put(allocationPath.getBatch().getId(), 0L);
+            }
+        }
+
+        if (schedule.getAllocationList() != null) {
```

Comment:
The level of nesting here is a bit too much to be understandable.
I suggest you either turn the blocks into their own functions, or (where applicable) invert the logic.
Example:

     for (...) {
         if (condition) {
             // The positive side.
         } else {
             // The negative side.
         }
     }

This can be turned into:

     for (...) {
         if (!condition) {
             // The negative side.
             continue;
         }
         // The positive side.
     }

See how this reduces the overall cognitive complexity?

Reply:
Sure. Will incorporate this change.